src/udp_socket.cpp

```diff
@@ -932,8 +935,19 @@ void socks5::connect2(error_code const& e)
 
 	if (atyp == 1)
 	{
-		m_udp_proxy_addr.address(address_v4(read_uint32(p)));
-		m_udp_proxy_addr.port(read_uint16(p));
+		ADD_OUTSTANDING_ASYNC("socks5::read_bindaddr");
+		m_tmp_buf[0] = *p;
+		boost::asio::async_read(m_socks5_sock, boost::asio::buffer(m_tmp_buf.data() + 1, 5)
+			, std::bind(&socks5::read_bindaddr, self(), _1));
+	}
+	else if (atyp == 3)
+	{
+		// we need to skip DOMAINNAME to imitate Python socks client
+		std::size_t const len = read_uint8(p);
```

Comment:
After this read `p` advances past the last read (5th) byte and points at the 6th byte which is the first hostname character.

Reply:
that's right. We get to this callback after having read 5 bytes, so the length (the 5th byte) is valid here. this is the last valid byte of the read buffer.

This was an optimization over your original patch where you would read up to the `ATYP` as the first read operation, instead I read up to the byte following `ATYP`. That way, if it's a hostname, I know how many more bytes to read exactly, and only have to issue a single additional read operation.